File: ios/CareLog/CareLog/FHIR/Services/LocalFHIRRepository.swift
//
//  LocalFHIRRepository.swift
//  CareLog
//
//  Repository for local FHIR resource storage using SwiftData
//

import Foundation
import SwiftData

/// Repository for local FHIR resource storage.
///
/// Provides offline-first CRUD operations with sync status tracking.
/// All data is stored locally first, then synced to the server
/// when connectivity is available.
@MainActor
class LocalFHIRRepository: ObservableObject {
    private let modelContainer: ModelContainer
    private let modelContext: ModelContext

    init() throws {
        let schema = Schema([
            LocalObservation.self,
            LocalDocumentReference.self
        ])

        let modelConfiguration = ModelConfiguration(
            schema: schema,
            isStoredInMemoryOnly: false,
            allowsSave: true
        )

        modelContainer = try ModelContainer(for: schema, configurations: [modelConfiguration])
        modelContext = modelContainer.mainContext
    }

    // MARK: - Observation Operations

    /// Save an observation locally.
    func saveObservation(_ observation: FHIRObservation) throws -> String {
        let localId = UUID().uuidString

        let localObservation = LocalObservation(
            localId: localId,
            serverId: observation.id,
            patientId: observation.patientId,
            type: observation.type,
            effectiveDateTime: observation.effectiveDateTime,
            value: observation.value,
            unit: observation.unit,
            systolicValue: observation.components?.first(where: { $0.type == .systolicBP })?.value,
            diastolicValue: observation.components?.first(where: { $0.type == .diastolicBP })?.value,
            interpretation: observation.interpretation,
            performerId: observation.performerId,
            performerType: observation.performerType,
            note: observation.note,
            isFasting: observation.context?.fasting ?? false,
            isPostMeal: observation.context?.postMeal ?? false,
            status: observation.status,
            syncStatus: observation.id != nil ? .synced : .pending
        )

        modelContext.insert(localObservation)
        try modelContext.save()

        return localId
    }

    /// Get an observation by local ID.
    func getObservation(localId: String) throws -> FHIRObservation? {
        let descriptor = FetchDescriptor<LocalObservation>(
            predicate: #Predicate { $0.localId == localId }
        )
        guard let local = try modelContext.fetch(descriptor).first else {
            return nil
        }
        return local.toFHIRObservation()
    }

    /// Get all observations for a patient.
    func getObservationsForPatient(_ patientId: String) throws -> [FHIRObservation] {
        let descriptor = FetchDescriptor<LocalObservation>(
            predicate: #Predicate { $0.patientId == patientId },
            sortBy: [SortDescriptor(\.effectiveDateTime, order: .reverse)]
        )
        return try modelContext.fetch(descriptor).map { $0.toFHIRObservation() }
    }

    /// Get observations by type for a patient.
    func getObservationsByType(_ patientId: String, type: ObservationType) throws -> [FHIRObservation] {
        let typeRawValue = type.rawValue
        let descriptor = FetchDescriptor<LocalObservation>(
            predicate: #Predicate {
                $0.patientId == patientId && $0.typeRawValue == typeRawValue
            },
            sortBy: [SortDescriptor(\.effectiveDateTime, order: .reverse)]
        )
        return try modelContext.fetch(descriptor).map { $0.toFHIRObservation() }
    }

    /// Get the latest observation of a type for a patient.
    func getLatestObservation(_ patientId: String, type: ObservationType) throws -> FHIRObservation? {
        let typeRawValue = type.rawValue
        var descriptor = FetchDescriptor<LocalObservation>(
            predicate: #Predicate {
                $0.patientId == patientId && $0.typeRawValue == typeRawValue
            },
            sortBy: [SortDescriptor(\.effectiveDateTime, order: .reverse)]
        )
        descriptor.fetchLimit = 1
        return try modelContext.fetch(descriptor).first?.toFHIRObservation()
    }

    /// Get observations pending sync.
    func getPendingSyncObservations(limit: Int = 50) throws -> [LocalObservation] {
        let pendingStatus = SyncStatus.pending.rawValue
        var descriptor = FetchDescriptor<LocalObservation>(
            predicate: #Predicate { $0.syncStatusRawValue == pendingStatus },
            sortBy: [SortDescriptor(\.createdAt)]
        )
        descriptor.fetchLimit = limit
        return try modelContext.fetch(descriptor)
    }

    /// Mark an observation as synced.
    func markObservationSynced(localId: String, serverId: String) throws {
        let descriptor = FetchDescriptor<LocalObservation>(
            predicate: #Predicate { $0.localId == localId }
        )
        guard let observation = try modelContext.fetch(descriptor).first else { return }

        observation.serverId = serverId
        observation.syncStatus = .synced
        observation.syncError = nil
        observation.updatedAt = Date()

        try modelContext.save()
    }

    /// Mark an observation sync as failed.
    func markObservationSyncFailed(localId: String, error: String) throws {
        let descriptor = FetchDescriptor<LocalObservation>(
            predicate: #Predicate { $0.localId == localId }
        )
        guard let observation = try modelContext.fetch(descriptor).first else { return }

        observation.syncStatus = .failed
        observation.syncError = error
        observation.syncAttempts += 1
        observation.lastSyncAttempt = Date()
        observation.updatedAt = Date()

        try modelContext.save()
    }

    /// Get count of observations pending sync.
    func getPendingObservationCount() throws -> Int {
        let pendingStatus = SyncStatus.pending.rawValue
        let failedStatus = SyncStatus.failed.rawValue
        let descriptor = FetchDescriptor<LocalObservation>(
            predicate: #Predicate {
                $0.syncStatusRawValue == pendingStatus || $0.syncStatusRawValue == failedStatus
            }
        )
        return try modelContext.fetchCount(descriptor)
    }

    /// Delete an observation.
    func deleteObservation(localId: String) throws {
        let descriptor = FetchDescriptor<LocalObservation>(
            predicate: #Predicate { $0.localId == localId }
        )
        if let observation = try modelContext.fetch(descriptor).first {
            modelContext.delete(observation)
            try modelContext.save()
        }
    }

    // MARK: - DocumentReference Operations

    /// Save a document reference locally.
    func saveDocumentReference(_ document: FHIRDocumentReference, localFilePath: String) throws -> String {
        let localId = UUID().uuidString

        let localDocument = LocalDocumentReference(
            localId: localId,
            serverId: document.id,
            patientId: document.patientId,
            documentId: document.documentId,
            type: document.type,
            title: document.title,
            description: document.description,
            localFilePath: localFilePath,
            contentUrl: document.contentUrl,
            contentType: document.contentType,
            size: document.size,
            date: document.date,
            authorId: document.authorId,
            authorName: document.authorName,
            authorType: document.authorType,
            documentStatus: document.status,
            fileUploadStatus: document.contentUrl != nil ? .synced : .pending,
            fhirSyncStatus: document.id != nil ? .synced : .pending
        )

        modelContext.insert(localDocument)
        try modelContext.save()

        return localId
    }

    /// Get documents for a patient.
    func getDocumentsForPatient(_ patientId: String) throws -> [FHIRDocumentReference] {
        let descriptor = FetchDescriptor<LocalDocumentReference>(
            predicate: #Predicate { $0.patientId == patientId },
            sortBy: [SortDescriptor(\.date, order: .reverse)]
        )
        return try modelContext.fetch(descriptor).map { $0.toFHIRDocumentReference() }
    }

    /// Get documents pending file upload.
    func getPendingFileUploads(limit: Int = 10) throws -> [LocalDocumentReference] {
        let pendingStatus = SyncStatus.pending.rawValue
        var descriptor = FetchDescriptor<LocalDocumentReference>(
            predicate: #Predicate { $0.fileUploadStatusRawValue == pendingStatus },
            sortBy: [SortDescriptor(\.createdAt)]
        )
        descriptor.fetchLimit = limit
        return try modelContext.fetch(descriptor)
    }

    /// Mark a document's file as uploaded.
    func markFileUploaded(localId: String, contentUrl: String) throws {
        let descriptor = FetchDescriptor<LocalDocumentReference>(
            predicate: #Predicate { $0.localId == localId }
        )
        guard let document = try modelContext.fetch(descriptor).first else { return }

        document.contentUrl = contentUrl
        document.fileUploadStatus = .synced
        document.syncError = nil
        document.updatedAt = Date()

        try modelContext.save()
    }

    /// Mark a document's file upload as failed.
    func markFileUploadFailed(localId: String, error: String) throws {
        let descriptor = FetchDescriptor<LocalDocumentReference>(
            predicate: #Predicate { $0.localId == localId }
        )
        guard let document = try modelContext.fetch(descriptor).first else { return }

        document.fileUploadStatus = .failed
        document.syncError = error
        document.syncAttempts += 1
        document.updatedAt = Date()

        try modelContext.save()
    }

    /// Get documents pending FHIR sync.
    func getPendingFHIRSync(limit: Int = 50) throws -> [LocalDocumentReference] {
        let syncedUpload = SyncStatus.synced.rawValue
        let pendingFhir = SyncStatus.pending.rawValue
        var descriptor = FetchDescriptor<LocalDocumentReference>(
            predicate: #Predicate {
                $0.fileUploadStatusRawValue == syncedUpload &&
                $0.fhirSyncStatusRawValue == pendingFhir
            },
            sortBy: [SortDescriptor(\.createdAt)]
        )
        descriptor.fetchLimit = limit
        return try modelContext.fetch(descriptor)
    }

    /// Mark a document as FHIR synced.
    func markDocumentFHIRSynced(localId: String, serverId: String) throws {
        let descriptor = FetchDescriptor<LocalDocumentReference>(
            predicate: #Predicate { $0.localId == localId }
        )
        guard let document = try modelContext.fetch(descriptor).first else { return }

        document.serverId = serverId
        document.fhirSyncStatus = .synced
        document.syncError = nil
        document.updatedAt = Date()

        try modelContext.save()
    }

    /// Delete a document reference.
    func deleteDocumentReference(localId: String) throws {
        let descriptor = FetchDescriptor<LocalDocumentReference>(
            predicate: #Predicate { $0.localId == localId }
        )
        if let document = try modelContext.fetch(descriptor).first {
            modelContext.delete(document)
            try modelContext.save()
        }
    }
}

// MARK: - Conversion Extensions

extension LocalObservation {
    func toFHIRObservation() -> FHIRObservation {
        let components: [ObservationComponent]?
        if type == .bloodPressure {
            var comps: [ObservationComponent] = []
            if let systolic = systolicValue {
                comps.append(ObservationComponent(type: .systolicBP, value: systolic, unit: "mmHg"))
            }
            if let diastolic = diastolicValue {
                comps.append(ObservationComponent(type: .diastolicBP, value: diastolic, unit: "mmHg"))
            }
            components = comps.isEmpty ? nil : comps
        } else {
            components = nil
        }

        return FHIRObservation(
            id: serverId,
            patientId: patientId,
            type: type,
            effectiveDateTime: effectiveDateTime,
            value: value,
            unit: unit,
            components: components,
            interpretation: interpretation,
            performerId: performerId,
            performerType: performerType,
            note: note,
            context: MeasurementContext(fasting: isFasting, postMeal: isPostMeal),
            status: status
        )
    }
}

extension LocalDocumentReference {
    func toFHIRDocumentReference() -> FHIRDocumentReference {
        FHIRDocumentReference(
            id: serverId,
            patientId: patientId,
            documentId: documentId,
            type: type,
            title: title,
            description: documentDescription,
            contentUrl: contentUrl ?? localFilePath,
            contentType: contentType,
            size: size,
            date: date,
            authorId: authorId,
            authorName: authorName,
            authorType: authorType,
            status: documentStatus
        )
    }
}

File: android/app/src/main/java/com/carelog/fhir/repository/LocalFhirRepository.kt
package com.carelog.fhir.repository

import com.carelog.fhir.client.DocumentType
import com.carelog.fhir.client.ObservationType
import com.carelog.fhir.local.dao.DocumentReferenceDao
import com.carelog.fhir.local.dao.ObservationDao
import com.carelog.fhir.local.entities.LocalDocumentReference
import com.carelog.fhir.local.entities.LocalObservation
import com.carelog.fhir.local.entities.SyncStatus
import com.carelog.fhir.models.*
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Repository for local FHIR resource storage.
 *
 * Provides offline-first CRUD operations with sync status tracking.
 * All data is stored locally first, then synced to the server
 * when connectivity is available.
 */
@Singleton
class LocalFhirRepository @Inject constructor(
    private val observationDao: ObservationDao,
    private val documentReferenceDao: DocumentReferenceDao
) {

    // ==================== Observation Operations ====================

    /**
     * Save an observation locally.
     * Returns the local ID for tracking.
     */
    suspend fun saveObservation(observation: FhirObservation): String {
        val localId = UUID.randomUUID().toString()
        val localObservation = observation.toLocalEntity(localId)
        observationDao.insert(localObservation)
        return localId
    }

    /**
     * Get an observation by local ID.
     */
    suspend fun getObservation(localId: String): FhirObservation? {
        return observationDao.getById(localId)?.toFhirObservation()
    }

    /**
     * Get all observations for a patient as a Flow.
     */
    fun getObservationsForPatient(patientId: String): Flow<List<FhirObservation>> {
        return observationDao.getByPatientId(patientId).map { list ->
            list.map { it.toFhirObservation() }
        }
    }

    /**
     * Get observations by type for a patient.
     */
    fun getObservationsByType(
        patientId: String,
        type: ObservationType
    ): Flow<List<FhirObservation>> {
        return observationDao.getByPatientIdAndType(patientId, type).map { list ->
            list.map { it.toFhirObservation() }
        }
    }

    /**
     * Get the latest observation of a type for a patient.
     */
    suspend fun getLatestObservation(
        patientId: String,
        type: ObservationType
    ): FhirObservation? {
        return observationDao.getLatestByType(patientId, type)?.toFhirObservation()
    }

    /**
     * Get observations pending sync.
     */
    suspend fun getPendingSyncObservations(limit: Int = 50): List<LocalObservation> {
        return observationDao.getPendingSync(SyncStatus.PENDING, limit)
    }

    /**
     * Mark an observation as synced.
     */
    suspend fun markObservationSynced(localId: String, serverId: String) {
        observationDao.markSynced(localId, serverId)
    }

    /**
     * Mark an observation sync as failed.
     */
    suspend fun markObservationSyncFailed(localId: String, error: String) {
        observationDao.markSyncFailed(localId, error)
    }

    /**
     * Get count of observations pending sync.
     */
    fun getPendingObservationCount(): Flow<Int> {
        return observationDao.countPendingSync()
    }

    /**
     * Delete an observation.
     */
    suspend fun deleteObservation(localId: String) {
        observationDao.deleteById(localId)
    }

    /**
     * Get all observations pending sync as a Flow.
     */
    fun getPendingObservations(): Flow<List<LocalObservation>> {
        return observationDao.getPendingSyncFlow(SyncStatus.PENDING)
    }

    /**
     * Get all observations for a patient.
     */
    fun getAllObservations(patientId: String): Flow<List<LocalObservation>> {
        return observationDao.getByPatientId(patientId)
    }

    /**
     * Get count of failed observation syncs.
     */
    fun getFailedObservationCount(): Flow<Int> {
        return observationDao.countByStatus(SyncStatus.FAILED)
    }

    /**
     * Update an observation's sync status.
     */
    suspend fun updateObservation(observation: LocalObservation) {
        observationDao.update(observation)
    }

    /**
     * Mark observation as failed with error.
     */
    suspend fun markObservationFailed(localId: String, error: String) {
        observationDao.markSyncFailed(localId, error)
    }

    // ==================== DocumentReference Operations ====================

    /**
     * Save a document reference locally.
     * Returns the local ID for tracking.
     */
    suspend fun saveDocumentReference(document: FhirDocumentReference, localFilePath: String): String {
        val localId = UUID.randomUUID().toString()
        val localDocument = document.toLocalEntity(localId, localFilePath)
        documentReferenceDao.insert(localDocument)
        return localId
    }

    /**
     * Get a document reference by local ID.
     */
    suspend fun getDocumentReference(localId: String): FhirDocumentReference? {
        return documentReferenceDao.getById(localId)?.toFhirDocumentReference()
    }

    /**
     * Get all documents for a patient as a Flow.
     */
    fun getDocumentsForPatient(patientId: String): Flow<List<FhirDocumentReference>> {
        return documentReferenceDao.getByPatientId(patientId).map { list ->
            list.map { it.toFhirDocumentReference() }
        }
    }

    /**
     * Get documents by type for a patient.
     */
    fun getDocumentsByType(
        patientId: String,
        type: DocumentType
    ): Flow<List<FhirDocumentReference>> {
        return documentReferenceDao.getByPatientIdAndType(patientId, type).map { list ->
            list.map { it.toFhirDocumentReference() }
        }
    }

    /**
     * Get documents pending file upload.
     */
    suspend fun getPendingFileUploads(limit: Int = 10): List<LocalDocumentReference> {
        return documentReferenceDao.getPendingFileUploads(SyncStatus.PENDING, limit)
    }

    /**
     * Mark a document's file as uploaded.
     */
    suspend fun markFileUploaded(localId: String, contentUrl: String) {
        documentReferenceDao.markFileUploaded(localId, contentUrl)
    }

    /**
     * Mark a document's file upload as failed.
     */
    suspend fun markFileUploadFailed(localId: String, error: String) {
        documentReferenceDao.markFileUploadFailed(localId, error)
    }

    /**
     * Get documents pending FHIR sync (after file upload).
     */
    suspend fun getPendingFhirSync(limit: Int = 50): List<LocalDocumentReference> {
        return documentReferenceDao.getPendingFhirSync(SyncStatus.PENDING, limit)
    }

    /**
     * Mark a document as FHIR synced.
     */
    suspend fun markDocumentFhirSynced(localId: String, serverId: String) {
        documentReferenceDao.markFhirSynced(localId, serverId)
    }

    /**
     * Mark a document's FHIR sync as failed.
     */
    suspend fun markDocumentFhirSyncFailed(localId: String, error: String) {
        documentReferenceDao.markFhirSyncFailed(localId, error)
    }

    /**
     * Get count of pending uploads + syncs.
     */
    fun getPendingDocumentCount(): Flow<Int> {
        return documentReferenceDao.countPendingFileUploads()
    }

    /**
     * Delete a document reference.
     */
    suspend fun deleteDocumentReference(localId: String) {
        documentReferenceDao.deleteById(localId)
    }

    /**
     * Get all documents pending sync as a Flow.
     */
    fun getPendingDocuments(): Flow<List<LocalDocumentReference>> {
        return documentReferenceDao.getPendingFhirSyncFlow(SyncStatus.PENDING)
    }

    /**
     * Get count of failed document syncs.
     */
    fun getFailedDocumentCount(): Flow<Int> {
        return documentReferenceDao.countByStatus(SyncStatus.FAILED)
    }

    /**
     * Update a document reference.
     */
    suspend fun updateDocumentReference(document: LocalDocumentReference) {
        documentReferenceDao.update(document)
    }

    /**
     * Mark document as synced.
     */
    suspend fun markDocumentSynced(localId: String, serverId: String) {
        documentReferenceDao.markFhirSynced(localId, serverId)
    }

    /**
     * Mark document sync as failed.
     */
    suspend fun markDocumentFailed(localId: String, error: String) {
        documentReferenceDao.markFhirSyncFailed(localId, error)
    }

    /**
     * Get last sync time (most recent synced observation).
     */
    fun getLastSyncTime(): Flow<Long?> {
        return observationDao.getLastSyncTime()
    }

    // ==================== Conversion Extensions ====================

    private fun FhirObservation.toLocalEntity(localId: String): LocalObservation {
        return LocalObservation(
            localId = localId,
            serverId = id,
            patientId = patientId,
            type = type,
            effectiveDateTime = java.time.Instant.parse(effectiveDateTime).epochSecond,
            value = value,
            unit = unit,
            systolicValue = components?.find { it.type == ObservationType.SYSTOLIC_BP }?.value,
            diastolicValue = components?.find { it.type == ObservationType.DIASTOLIC_BP }?.value,
            interpretation = interpretation,
            performerId = performerId,
            performerType = performerType,
            note = note,
            isFasting = context?.fasting ?: false,
            isPostMeal = context?.postMeal ?: false,
            status = status,
            syncStatus = if (id != null) SyncStatus.SYNCED else SyncStatus.PENDING
        )
    }

    private fun LocalObservation.toFhirObservation(): FhirObservation {
        val components = if (type == ObservationType.BLOOD_PRESSURE) {
            listOfNotNull(
                systolicValue?.let { ObservationComponent(ObservationType.SYSTOLIC_BP, it, "mmHg") },
                diastolicValue?.let { ObservationComponent(ObservationType.DIASTOLIC_BP, it, "mmHg") }
            ).takeIf { it.isNotEmpty() }
        } else null

        return FhirObservation(
            id = serverId,
            patientId = patientId,
            type = type,
            effectiveDateTime = java.time.Instant.ofEpochSecond(effectiveDateTime)
                .atZone(java.time.ZoneOffset.UTC)
                .format(java.time.format.DateTimeFormatter.ISO_INSTANT),
            value = value,
            unit = unit,
            components = components,
            interpretation = interpretation,
            performerId = performerId,
            performerType = performerType,
            note = note,
            context = MeasurementContext(fasting = isFasting, postMeal = isPostMeal),
            status = status
        )
    }

    private fun FhirDocumentReference.toLocalEntity(
        localId: String,
        localFilePath: String
    ): LocalDocumentReference {
        return LocalDocumentReference(
            localId = localId,
            serverId = id,
            patientId = patientId,
            documentId = documentId,
            type = type,
            title = title,
            description = description,
            localFilePath = localFilePath,
            contentUrl = contentUrl,
            contentType = contentType,
            size = size,
            date = java.time.Instant.parse(date).toEpochMilli(),
            authorId = authorId,
            authorName = authorName,
            authorType = authorType,
            status = status,
            fileUploadStatus = if (contentUrl != null) SyncStatus.SYNCED else SyncStatus.PENDING,
            fhirSyncStatus = if (id != null) SyncStatus.SYNCED else SyncStatus.PENDING
        )
    }

    private fun LocalDocumentReference.toFhirDocumentReference(): FhirDocumentReference {
        return FhirDocumentReference(
            id = serverId,
            patientId = patientId,
            documentId = documentId,
            type = type,
            title = title,
            description = description,
            contentUrl = contentUrl ?: localFilePath,
            contentType = contentType,
            size = size,
            date = java.time.Instant.ofEpochMilli(date).toString(),
            authorId = authorId,
            authorName = authorName,
            authorType = authorType,
            status = status
        )
    }
}
